here-naksha-lib-handlers/src/main/java/com/here/naksha/lib/handlers/DefaultStorageHandler.java

```diff
@@ -113,164 +92,197 @@ protected EventProcessingStrategy processingStrategyFor(IEvent event) {
+      final @NotNull NakshaCollection collection,
+      final @NotNull WriteRequest wf,
+      final OperationAttempt operationAttempt)
+          throws Throwable {
     logger.info("Processing WriteFeatures against {}", collection.getId());
-    return forwardWriteRequest(
-        ctx,
-        storageImpl,
-        wf,
-        re -> reattemptFeatureRequest(ctx, storageImpl, collection, wf, operationAttempt, re));
+    return forwardWriteRequest(ctx, storageImpl, wf, re -> {
+      try {
+        return reattemptFeatureRequest(ctx, storageImpl, collection, wf, operationAttempt, re);
+      } catch (Throwable e) {
+        return new ErrorResponse(
+                NakshaError.EXCEPTION, "Error even after retrying writing features: ", null, e);
+      }
+    });
   }
 
-  private @NotNull Result forwardWriteCollections(
+  private @NotNull Response forwardWriteCollections(
       final @NotNull NakshaContext ctx,
       final @NotNull IStorage storageImpl,
-      final @NotNull XyzCollection collection,
-      final @NotNull WriteCollections<?, ?, ?> wc,
-      final OperationAttempt operationAttempt) {
+      final @NotNull NakshaCollection collection,
+      final @NotNull WriteRequest wc,
+      final OperationAttempt operationAttempt)
+          throws Throwable {
     logger.info("Processing WriteCollections against {}", collection.getId());
     if (isUpdateCollectionRequest(wc)) {
       if (properties.getAutoCreateCollection()) {
-        return forwardWriteRequest(
-            ctx,
-            storageImpl,
-            wc,
-            re -> reattemptCollectionRequest(ctx, storageImpl, collection, wc, operationAttempt, re));
+        return forwardWriteRequest(ctx, storageImpl, wc, re -> {
+          try {
+            return reattemptCollectionRequest(ctx, storageImpl, collection, wc, operationAttempt, re);
+          } catch (Throwable e) {
+            return new ErrorResponse(
+                    NakshaError.EXCEPTION, "Error even after retrying updating collection: ", null, e);
+          }
+        });
       } else {
         logger.info(
             "Received update collection request but autoCreate is not enabled, returning success without any action");
-        return new SuccessResult();
+        return new SuccessResponse();
       }
     } else if (isPurgeCollectionRequest(wc)) {
       if (properties.getAutoDeleteCollection()) {
-        return forwardWriteRequest(
-            ctx,
-            storageImpl,
-            wc,
-            re -> reattemptCollectionRequest(ctx, storageImpl, collection, wc, operationAttempt, re));
+        return forwardWriteRequest(ctx, storageImpl, wc, re -> {
+          try {
+            return reattemptCollectionRequest(ctx, storageImpl, collection, wc, operationAttempt, re);
+          } catch (Throwable e) {
+            return new ErrorResponse(
+                    NakshaError.EXCEPTION, "Error even after retrying purging collection: ", null, e);
+          }
```

Comment:
This, and a lot of other throwable-related changes, are IMO not required and they obscure a code - the changes that came with v3 shouldn't require this much of a change.
I feel like most of the `try...catch` statements as well as `(Throwable) ...` casts are not necessary and might even affect the logic (retries to be specific).

I think it'd be best if we discussed this class on call